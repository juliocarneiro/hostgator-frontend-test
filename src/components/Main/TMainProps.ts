export type TMainProducts = {
  name: string
  cycle: [
    {
      annually: { priceOrder: number; months: number }
      triennially: { priceOrder: number; months: number }
      monthly: { priceOrder: number; months: number }
    }
  ]
  id: string
}

export type TMainSelectedTime = string
