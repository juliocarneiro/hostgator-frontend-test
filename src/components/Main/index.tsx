import { useEffect, useState } from 'react'
import getProducts from 'utils'

import Header from '../Header'
import Filter from '../Filter'
import Plans from '../Plans'
import Loader from '../Loader'

type Products = {
  name: string
  cycle: [
    {
      annually: { priceOrder: number; months: number }
      triennially: { priceOrder: number; months: number }
      monthly: { priceOrder: number; months: number }
    }
  ]
  id: string
}

type SelectedTime = string

const Main = () => {
  const [productsData, setProductsData] = useState<Products[] | []>([])
  const [selectedTime, setSelectedTime] = useState<SelectedTime>('three_years')
  const [isLoading, setIsLoading] = useState<boolean>(true)

  useEffect(() => {
    async function fetchData() {
      const { products } = await getProducts()

      const productsFilter = products.filter(
        (e: { name: string }) =>
          e.name === 'Plano P' ||
          e.name === 'Plano M' ||
          e.name === 'Plano Turbo'
      )

      setProductsData(productsFilter)
      setIsLoading(false)
    }
    fetchData()
  }, [])

  useEffect(() => {
    setSelectedTime(selectedTime)
  }, [selectedTime])

  return (
    <>
      {isLoading ? (
        <Loader isLoading={isLoading} />
      ) : (
        <div className="animate__animated animate__fadeIn">
          <Header />
          <Filter
            selectedTime={selectedTime}
            setSelectedTime={setSelectedTime}
          />
          <Plans selectedTime={selectedTime} productsData={productsData} />
        </div>
      )}
    </>
  )
}

export default Main
