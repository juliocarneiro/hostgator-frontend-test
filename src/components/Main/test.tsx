import { renderHook } from '@testing-library/react-hooks'
import Main from '.'

describe('should render Main', () => {
  test('<Main /> to be visible', async () => {
    const { result, waitForNextUpdate } = renderHook(() => Main())
    await waitForNextUpdate()
    expect(result).toMatchSnapshot()
  })
})
