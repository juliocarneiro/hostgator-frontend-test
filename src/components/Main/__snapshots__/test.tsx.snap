// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Main <Main /> to be visible 1`] = `
Object {
  "all": Array [
    <React.Fragment>
      <Loader
        isLoading={true}
      />
    </React.Fragment>,
    <React.Fragment>
      <Loader
        isLoading={true}
      />
    </React.Fragment>,
    <React.Fragment>
      <div
        className="animate__animated animate__fadeIn"
      >
        <Header />
        <Filter
          selectedTime="three_years"
          setSelectedTime={[Function]}
        />
        <Plans
          productsData={
            Array [
              Object {
                "cycle": Array [
                  Object {
                    "annually": Object {
                      "months": 12,
                      "priceOrder": "220.66",
                      "priceRenew": "220.66",
                    },
                    "biennially": Object {
                      "months": 24,
                      "priceOrder": "393.36",
                      "priceRenew": "393.36",
                    },
                    "monthly": Object {
                      "months": 1,
                      "priceOrder": "24.19",
                      "priceRenew": "24.19",
                    },
                    "quarterly": Object {
                      "months": 3,
                      "priceOrder": "67.17",
                      "priceRenew": "67.17",
                    },
                    "semiannually": Object {
                      "months": 6,
                      "priceOrder": "128.34",
                      "priceRenew": "128.34",
                    },
                    "triennially": Object {
                      "months": 36,
                      "priceOrder": "561.13",
                      "priceRenew": "561.13",
                    },
                  },
                ],
                "id": 5,
                "name": "Plano P",
              },
              Object {
                "cycle": Array [
                  Object {
                    "annually": Object {
                      "months": 12,
                      "priceOrder": "286.66",
                      "priceRenew": "286.66",
                    },
                    "biennially": Object {
                      "months": 24,
                      "priceOrder": "532.56",
                      "priceRenew": "532.56",
                    },
                    "monthly": Object {
                      "months": 1,
                      "priceOrder": "29.69",
                      "priceRenew": "29.69",
                    },
                    "quarterly": Object {
                      "months": 3,
                      "priceOrder": "82.77",
                      "priceRenew": "82.77",
                    },
                    "semiannually": Object {
                      "months": 6,
                      "priceOrder": "159.54",
                      "priceRenew": "159.54",
                    },
                    "triennially": Object {
                      "months": 36,
                      "priceOrder": "764.22",
                      "priceRenew": "764.22",
                    },
                  },
                ],
                "id": 6,
                "name": "Plano M",
              },
              Object {
                "cycle": Array [
                  Object {
                    "annually": Object {
                      "months": 12,
                      "priceOrder": "529.07",
                      "priceRenew": "529.07",
                    },
                    "biennially": Object {
                      "months": 24,
                      "priceOrder": "1032.95",
                      "priceRenew": "1032.95",
                    },
                    "monthly": Object {
                      "months": 1,
                      "priceOrder": "47.24",
                      "priceRenew": "47.24",
                    },
                    "quarterly": Object {
                      "months": 3,
                      "priceOrder": "138.57",
                      "priceRenew": "138.57",
                    },
                    "semiannually": Object {
                      "months": 6,
                      "priceOrder": "270.84",
                      "priceRenew": "270.84",
                    },
                    "triennially": Object {
                      "months": 36,
                      "priceOrder": "1511.62",
                      "priceRenew": "1511.62",
                    },
                  },
                ],
                "id": 335,
                "name": "Plano Turbo",
              },
            ]
          }
          selectedTime="three_years"
        />
      </div>
    </React.Fragment>,
  ],
  "current": <React.Fragment>
    <div
      className="animate__animated animate__fadeIn"
    >
      <Header />
      <Filter
        selectedTime="three_years"
        setSelectedTime={[Function]}
      />
      <Plans
        productsData={
          Array [
            Object {
              "cycle": Array [
                Object {
                  "annually": Object {
                    "months": 12,
                    "priceOrder": "220.66",
                    "priceRenew": "220.66",
                  },
                  "biennially": Object {
                    "months": 24,
                    "priceOrder": "393.36",
                    "priceRenew": "393.36",
                  },
                  "monthly": Object {
                    "months": 1,
                    "priceOrder": "24.19",
                    "priceRenew": "24.19",
                  },
                  "quarterly": Object {
                    "months": 3,
                    "priceOrder": "67.17",
                    "priceRenew": "67.17",
                  },
                  "semiannually": Object {
                    "months": 6,
                    "priceOrder": "128.34",
                    "priceRenew": "128.34",
                  },
                  "triennially": Object {
                    "months": 36,
                    "priceOrder": "561.13",
                    "priceRenew": "561.13",
                  },
                },
              ],
              "id": 5,
              "name": "Plano P",
            },
            Object {
              "cycle": Array [
                Object {
                  "annually": Object {
                    "months": 12,
                    "priceOrder": "286.66",
                    "priceRenew": "286.66",
                  },
                  "biennially": Object {
                    "months": 24,
                    "priceOrder": "532.56",
                    "priceRenew": "532.56",
                  },
                  "monthly": Object {
                    "months": 1,
                    "priceOrder": "29.69",
                    "priceRenew": "29.69",
                  },
                  "quarterly": Object {
                    "months": 3,
                    "priceOrder": "82.77",
                    "priceRenew": "82.77",
                  },
                  "semiannually": Object {
                    "months": 6,
                    "priceOrder": "159.54",
                    "priceRenew": "159.54",
                  },
                  "triennially": Object {
                    "months": 36,
                    "priceOrder": "764.22",
                    "priceRenew": "764.22",
                  },
                },
              ],
              "id": 6,
              "name": "Plano M",
            },
            Object {
              "cycle": Array [
                Object {
                  "annually": Object {
                    "months": 12,
                    "priceOrder": "529.07",
                    "priceRenew": "529.07",
                  },
                  "biennially": Object {
                    "months": 24,
                    "priceOrder": "1032.95",
                    "priceRenew": "1032.95",
                  },
                  "monthly": Object {
                    "months": 1,
                    "priceOrder": "47.24",
                    "priceRenew": "47.24",
                  },
                  "quarterly": Object {
                    "months": 3,
                    "priceOrder": "138.57",
                    "priceRenew": "138.57",
                  },
                  "semiannually": Object {
                    "months": 6,
                    "priceOrder": "270.84",
                    "priceRenew": "270.84",
                  },
                  "triennially": Object {
                    "months": 36,
                    "priceOrder": "1511.62",
                    "priceRenew": "1511.62",
                  },
                },
              ],
              "id": 335,
              "name": "Plano Turbo",
            },
          ]
        }
        selectedTime="three_years"
      />
    </div>
  </React.Fragment>,
  "error": undefined,
}
`;
